packages/python/accern_xyme/util.py

```diff
@@ -573,3 +574,47 @@ def maybe_json_loads(value: str) -> Any:
         return json.loads(escape_str(value))
     except json.JSONDecodeError:
         return None
+
+
+def compute_parallel(
+        tasks: List[int],
+        computing_fn: Callable[[int, threading.RLock], CT],
+        progress_fn: Optional[Callable[..., None]],
+        max_threads: int) -> List[Optional[CT]]:
+    a_tasks = list(enumerate(tasks))
+    task_count = 0
+    results: List[Optional[CT]] = [None] * len(a_tasks)
+    progress_lock = threading.RLock()
+    compute_lock = threading.RLock()
+    exc: Optional[BaseException] = None
+
+    def runner() -> None:
+        nonlocal task_count, exc
+
+        while True:
+            try:
+                cur_task = a_tasks.pop(0) if a_tasks else None
+                if cur_task is None or exc is not None:
+                    break
+                task_ix, offset = cur_task
+                results[task_ix] = computing_fn(offset, compute_lock)
+                task_count += 1
+                if progress_fn is not None:
+                    with progress_lock:
+                        progress_fn(task_count / len(tasks), False)
+            except BaseException as e:  # pylint: disable=broad-except
+                exc = e
+
+    ths = [
+        threading.Thread(target=runner, name=f"thread id: {tid}", daemon=True)
+        for tid in range(max_threads)
+    ]
+    for th in ths:
+        th.start()
+    for th in ths:
+        th.join()
+    if isinstance(exc, BaseException):
+        raise exc  # pylint: disable=raising-bad-type
```

Comment:
why is it bad type?

Reply:
also reraise the exception -- don't just raise the exception -- otherwise you lose the trace